### ISSUE TITLE
Issue with clink.register_match_generator(tilde_match, 1)

### ISSUE THREAD

**Issue by alastairUK** [cmderdev/cmder#2505]
### Purpose of the issue
- [ x] Bug report (encountered problems/errors)
- [ ] Feature request (request for new functionality)
- [ ] Question

### Version Information

1.3.17.1082 19.10.12.0

Windows 10 Pro 19042.630
### Description of the issue

I am using a script similar to the one here: https://github.com/cmderdev/cmder/issues/1051 and all was working fine in 1.3.16 but on upgrading to 1.3.17 nothing seems to happen if I type ~ and press tab. Has something changed to affect this?

**Comment by chrisant996** [comment]
What version of Clink does 1.3.17 use?  I think Cmder is currently deploying an older issue of Clink that has a regression in the tilde stuff, which has been fixed in newer Clink versions for a couple months or so.

What happens if you copy the latest version of Clink into the cmder `vendor\clink` directory?

**Comment by alastairUK** [comment]
Not using any extra .lua scripts just the standard installs:

1.3.17.1082 (from https://cmder.net/)  installs clink version 1.1.10.0 which doesn't handle ~ 

If I install cmder 1.3.16 from https://github.com/cmderdev/cmder/releases/tag/v1.3.16 this uses clink 0.4.9.45822 and all works fine with ~

If I download the latest clink from https://github.com/chrisant996/clink/releases/tag/v1.1.34 (is this the right location as the clink from https://github.com/mridgers/clink/tags seems < 1.1.10.0)

clink --version
1.1.34.d161e9

~ still doesn't seem to work.

**Comment by chrisant996** [comment]
Yes, though I would have suggested v1.1.44, the latest release of Clink, from here: https://github.com/chrisant996/clink/releases.

When I use Clink by itself, without Cmder, then tilde completion works as intended with both `menu-complete` and `complete`:
- `~\` or `~\a` (etc) completes.
- `~` by itself does not complete -- this is intentional:
    - There is no path separator, so it might be part of a file name, not a path.
    - I think this is probably why Cmder tries to replace Clink's tilde completion.

When I use Clink v1.1.x inside Cmder, then tilde completion has several problems, and it doesn't work at all with `menu-complete`.
This is because Cmder registers a custom `tilde_match` generator from Cmder's `vendor\clink.lua` file (which is not part of Clink).
Clink v1.0.0 changed how match generators work, and Clink v1.1.x added backward compatibility support to let existing scripts continue to work.

Unfortunately the backward compatibility support doesn't work for this particular script.

I'll see what I can do about that, because some other match generators might also need stronger backward compatibility support.

**Comment by daxgames** [comment]
We added the tilde matxh because it never worked with pre clink 1.x.x versions.  It could probably be removed from cmders ckink.lua.

**Comment by chrisant996** [comment]
Tilde match works fine in Clink 0.4.9, but just not if you _only_ type a `~`, and that's on purpose.

1. Launch Clink 0.4.9 standalone (not via Cmder).
2. Type `~` and try to complete -- nothing happens because it's not a directory yet, it's only a filename fragment.
3. Add `\` and try to complete -- now it completes because the path separator removed the ambiguity: `~\` is clearly a directory.

**Comment by chrisant996** [comment]
Also, the more I look at the implementation of `tilde_match`, it has some issues that Clink can't/shouldn't compensate for:

- It uses `%USERPROFILE%` but should use `%HOME%` (another place in Cmder can initialize HOME to something other than USERPROFILE, and then `tilde_match` will behave inconsistently from the rest of Cmder and Clink.
- It adds a bogus match with an appended `+` sign to try to trick Readline into certain behaviors.  But that bogus match breaks `menu-complete` and `clink-shim-menu-complete`.  I think `clink.suppress_char_append()` might achieve the same effect without adding a bogus match, but there's still the problem that the match generator doesn't actually generate the matches.
- It is implemented as a generator, but it doesn't generate the actual matches.

The backward compatibility issue is that in 0.4.9 match generators were given the partial word to be completed, but in v1.0.0 and higher match generators are no longer responsible for filtering.  This gains some performance benefits, among other things.

I'll still see what I can do to maybe pass the word to be completed and improve backward compatibility, but I'm not sure how feasible that will be.

**Comment by chrisant996** [comment]
I tried to make a lua function that would have the same/similar effect in new-Clink.  It isn't possible with the APIs currently available, because Clink caches the generated matches and expects generators to keep generating the same set of matches each time.  So `~d`<kbd>Tab</kbd><kbd>Backspace</kbd><kbd>Tab</kbd> inserts `C:\Users\yournamed` instead of `C:\users\yourname`.

I will look into:
1. Adding `rl.setvariable(name, value)` which Cmder could use to force `expand-tilde` on regardless of the inputrc file.  This is a little concerning, though:  it would make it possible for any script to change any Readline config variables whenever they want, which has potential to create confusion and inconsistent/intermittent issues if some script misuses it, or wants to temporarily change a variable but misses restoring the original value, etc.
2. Adding some kind of `match_builder:dontcachematches()` API that a generator can use to force a full generate pass next time a completion function is used.  This exists internally, but has not yet been exposed to lua scripts.  I need to dream up a good name for it; I'm not happy with the name above because it's phrased in terms of the internal implementation, and I prefer the name to be phrased in terms of the intent or role without any reference to implementation.

**Comment by chrisant996** [comment]
UPDATE:

Since `expand-tilde` is on by default in new-Clink, the simplest thing to do is just disable the `tilde_match` generator with new-Clink.  The easiest way to do that is wrap it in an `if` like the following.

Then Cmder automatically gets tilde expansion by default, with the broadest compatibility.  There are two drawbacks in Clink v1.1.45 and lower:  completing `~` or `~d` have no effect.  It's because Clink is trying to support files whose names start with `~`.  Bash doesn't care about that, because on *nix you wouldn't make a file whose name begins with tilde because tilde has special meaning in *nix.  But on Windows tilde does not have any special meaning, and sometimes people do create file names that start with tilde.  Clink is trying to cater to Windows, but maybe it should act closer to bash and interpret `~` as `%HOME%\` instead of `%HOME%`?  I'll look into possibly changing that detail.

#### vendor\clink.lua:
```lua
-- insert the set_prompt at the very beginning so that it runs first
if not clink.version_encoded then
    clink.register_match_generator(tilde_match, 1)
end
clink.prompt.register_filter(set_prompt_filter, 1)
```

* I said it's intentional that Clink does not complete `~` by itself.  But I acknowledge it's debatable.  For example, it's not how bash works: bash always treats a leading `~` as though it is `~/`, and it's impossible for completion of `~` to see any local `~foo` file in the current directory.  Since bash already has that quirk, it could be reas

**Comment by daxgames** [comment]
I just removed `tilde_match` from `%cmder_root\vendor\clink.lua` and released Cmder 1.3.18 with Clink 1.1.45.
